Mathlib/Computability/Primrec.lean

```diff
@@ -1155,6 +1158,125 @@ theorem nat_strong_rec (f : α → ℕ → σ) {g : α → List σ → Option σ
       simp [IH, H, List.range_succ]
 #align primrec.nat_strong_rec Primrec.nat_strong_rec
 
+theorem option_toList : Primrec (Option.toList : Option α → List α) :=
+  (option_casesOn Primrec.id (const [])
+    ((list_cons.comp Primrec.id (const [])).comp₂ Primrec₂.right)).of_eq
+  (fun o => by rcases o <;> simp)
+
+theorem list_lookup [DecidableEq α] : Primrec₂ (List.lookup : α → List (α × β) → Option β) :=
+  (to₂ <| list_rec snd (const none) <|
+    to₂ <|
+      cond (Primrec.beq.comp (fst.comp fst) (fst.comp $ fst.comp snd))
+        (option_some.comp $ snd.comp $ fst.comp snd)
+        (snd.comp $ snd.comp snd)).of_eq
+  fun a ps => by
+  induction' ps with p ps ih <;> simp[List.lookup, *]
+  cases ha : a == p.1 <;> simp[ha]; simpa using ih
+
+section nat_omega_rec
+
+section
+
+variable [DecidableEq β]
+
+private def mapGraph (M : List (β × γ)) (bs : List β) : List γ :=
+  bs.bind (Option.toList <| M.lookup ·)
+
+private def bindItr (l : β → List β) (b : β) : ℕ → List β := fun n =>
+  n.rec [b] (fun _ bs => bs.bind l)
+
+private def graph (m : β → ℕ) (l : β → List β) (g : β → List σ → Option σ) (b : β) :
```

Comment:
Can all of these `private def` and `private lemma` be moved into `nat_omega_rec'` as `let` and `have` respectively?